packages/schemas/src/text/uiRender.ts

```diff
@@ -0,0 +1,153 @@
+import type * as CSS from 'csstype';
+import { UIRenderProps, Schema, getDefaultFont } from '@pdfme/common';
+import type { TextSchema } from './types';
+import {
+  DEFAULT_FONT_SIZE,
+  DEFAULT_ALIGNMENT,
+  VERTICAL_ALIGN_TOP,
+  VERTICAL_ALIGN_MIDDLE,
+  VERTICAL_ALIGN_BOTTOM,
+  DEFAULT_VERTICAL_ALIGNMENT,
+  DEFAULT_LINE_HEIGHT,
+  DEFAULT_CHARACTER_SPACING,
+  DEFAULT_FONT_COLOR,
+} from './constants';
+import {
+  calculateDynamicFontSize,
+  getFontKitFont,
+  getBrowserVerticalFontAdjustments,
+} from './helper';
+
+const mapVerticalAlignToFlex = (verticalAlignmentValue: string | undefined) => {
+  switch (verticalAlignmentValue) {
+    case VERTICAL_ALIGN_TOP:
+      return 'flex-start';
+    case VERTICAL_ALIGN_MIDDLE:
+      return 'center';
+    case VERTICAL_ALIGN_BOTTOM:
+      return 'flex-end';
+  }
+  return 'flex-start';
+};
+
+export const uiRender = async (arg: UIRenderProps<TextSchema>) => {
+  const {
+    value,
+    schema,
+    rootElement,
+    mode,
+    onChange,
+    stopEditing,
+    tabIndex,
+    placeholder,
+    options,
+  } = arg;
+  const font = options?.font || getDefaultFont();
+
+  let dynamicFontSize: undefined | number = undefined;
+  if (schema.dynamicFontSize && value) {
+    dynamicFontSize = await calculateDynamicFontSize({
+      textSchema: schema,
+      font,
+      value,
+      startingFontSize: dynamicFontSize,
+    });
+  }
+
+  const fontKitFont = await getFontKitFont(schema, font);
+  // Depending on vertical alignment, we need to move the top or bottom of the font to keep
+  // it within it's defined box and align it with the generated pdf.
+  const { topAdj, bottomAdj } = getBrowserVerticalFontAdjustments(
+    fontKitFont,
+    dynamicFontSize ?? schema.fontSize ?? DEFAULT_FONT_SIZE,
+    schema.lineHeight ?? DEFAULT_LINE_HEIGHT,
+    schema.verticalAlignment ?? DEFAULT_VERTICAL_ALIGNMENT
+  );
+
+  const topAdjustment = topAdj;
+  const bottomAdjustment = bottomAdj;
+
+  const container = document.createElement('div');
+  function getBackgroundColor(mode: 'form' | 'viewer', value: string, schema: Schema) {
+    if (mode === 'form' && value && schema.backgroundColor) {
+      return schema.backgroundColor as string;
+    } else if (mode === 'viewer') {
+      return (schema.backgroundColor as string) ?? 'transparent';
+    } else {
+      return 'rgb(242 244 255 / 75%)';
+    }
+  }
+
+  const containerStyle: CSS.Properties = {
+    padding: 0,
+    resize: 'none',
+    backgroundColor: getBackgroundColor(mode, value, schema),
+    border: 'none',
+    display: 'flex',
+    flexDirection: 'column',
+    justifyContent: mapVerticalAlignToFlex(schema.verticalAlignment),
+    width: '100%',
+    height: '100%',
+  };
+  Object.assign(container.style, containerStyle);
+  rootElement.innerHTML = '';
```

Comment:
I see these files have changed from JSX, which feels a bit harder to read. Was it not possible to use JSX?

Reply:
@peteward 
Indeed, JSX is more readable.

However, requiring users to write plugins in React would mean forcing them to unify the UI package and React versions, as well as pre-transpile JSX.

While React might not be an option, handling UI rendering through direct DOM manipulation is a clear responsibility. Overall, this approach is the simplest.